src/passes/GlobalTypeOptimization.cpp

```diff
@@ -203,13 +203,19 @@ struct GlobalTypeOptimization : public Pass {
         // visibility, so do that here: we can only become immutable if the
         // parent can as well.
         auto super = type.getDeclaredSuperType();
-        if (super && !canBecomeImmutable.count(*super)) {
-          // No entry in canBecomeImmutable means nothing in the parent can
-          // become immutable. We don't need to check the specific field index,
-          // because visibility affects them all equally (i.e., if it is public
-          // then no field can be changed, and if it is private then this field
-          // can be changed, and perhaps more).
-          continue;
+        if (super) {
+          // The super may not contain the field, which is fine, so only check
+          // here if the field does exist in both.
+          if (i < super->getStruct().fields.size()) {
+            // No entry in canBecomeImmutable means nothing in the parent can
+            // become immutable, so check for both that and for an entry with
+            // "false".
+            if (!canBecomeImmutable.count(*super) ||
+                i >= canBecomeImmutable[*super].size() ||
```

Comment:
Let's use `canBecomeImmutable.find(*super)` to avoid the triple(!) lookup.

Reply:
Done.